sigbak.1
.\" Copyright (c) 2019 Tim van der Molen <tim@kariliq.nl>
.\"
.\" Permission to use, copy, modify, and distribute this software for any
.\" purpose with or without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
.\" WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.\" ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
.\" ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
.\" OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.Dd July 3, 2024
.Dt SIGBAK 1
.Os
.Sh NAME
.Nm sigbak
.Nd export messages from Signal Android backups
.Sh SYNOPSIS
.Nm sigbak
.Ar command
.Op Ar argument ...
.Sh DESCRIPTION
.Nm
is a utility to read the backups created by the Signal Android app.
It can be used to export messages, attachments and other data.
.Pp
A Signal backup consists primarily of an SQLite database.
This database contains the messages, amongst other things.
Attachments, avatars and stickers, however, are stored as separate files in the
backup.
.Pp
Every Signal backup is encrypted with a 30-digit passphrase.
By default,
.Nm
prompts for the passphrase.
Alternatively, the
.Fl p
option (see below) may be used to specify a file that contains the passphrase.
If the specified file is
.Sq - ,
the passphrase is read from standard input.
Whitespace in the passphrase is ignored.
.Pp
The commands are as follows:
.Bl -tag -width Ds
.Tg check
.It Ic check-backup Oo Fl p Ar passfile Oc Ar backup
.D1 Pq Alias: Ic check
.Pp
Check that the file
.Ar backup
can be decrypted and parsed correctly.
The check is aborted after the first encountered error.
.Tg dump
.It Ic dump-backup Oo Fl p Ar passfile Oc Ar backup
.D1 Pq Alias: Ic dump
.Pp
Print the raw contents of the file
.Ar backup
in a readable format on standard output.
.Tg att
.It Xo
.Ic export-attachments
.Op Fl aIMm
.Oo Fl p Ar passfile Oc
.Ar backup Op Ar directory
.Xc
.D1 Pq Alias: Ic att
.Pp
Export the attachments from the file
.Ar backup .
The attachment files are created in separate directories, one for each
conversation.
These directories are created in
.Ar directory ,
or in the current directory if
.Ar directory
is not specified.
.Pp
By default, long-message attachments are not exported.
Long-message attachments contain the text of longer messages.
The
.Fl a
option may be used to export these attachments, too.
.Pp
If the
.Fl I
option is specified, the attachment id is prepended to the filename of each
exported attachment.
.Pp
If the
.Fl M
option is specified, the file modification time of each exported attachment is
set to the time the attachment was sent.
The
.Fl m
option is similar, but uses the time the attachment was received.
.Tg avt
.It Ic export-avatars Oo Fl p Ar passfile Oc Ar backup Op Ar directory
.D1 Pq Alias: Ic avt
.Pp
Export all avatars from the file
.Ar backup
to
.Ar directory ,
or to the current directory if
.Ar directory
is not specified.
.Tg db
.It Ic export-database Oo Fl p Ar passfile Oc Ar backup Ar database
.D1 Pq Alias: Ic db
.Pp
Export the SQLite database from the file
.Ar backup
to the file
.Ar database .
.Tg msg
.It Xo
.Ic export-messages
.Oo Fl f Ar format Oc
.Oo Fl p Ar passfile Oc
.Ar backup Op Ar directory
.Xc
.D1 Pq Alias: Ic msg
.Pp
Export the messages from the file
.Ar backup .
The messages are written to separate files, one for each conversation.
These files are created in
.Ar directory ,
or in the current directory if
.Ar directory
is not specified.
.Pp
The
.Fl f
option may be used to specify the output format.
The following output formats are supported:
.Bl -tag -width Ds
.It Cm csv
Messages are written in CSV (comma-separated values) format.
See the
.Sx CSV FORMAT
section below for details.
.It Cm text
Messages are written as plain text.
This is the default.
.It Cm text-short
Messages are written as plain text, in short form.
Every message is written on a single line.
.El
.Tg stk
.It Ic export-stickers Oo Fl p Ar passfile Oc Ar backup Op Ar directory
.D1 Pq Alias: Ic stk
.Pp
Export all stickers from the file
.Ar backup
to
.Ar directory ,
or to the current directory if
.Ar directory
is not specified.
.El
.Sh CSV FORMAT
The
.Ic export-messages
command can export messages in CSV format.
In this format, each record describes either a message or a message reaction,
and consists of the following eight fields.
.Pp
The
.Em first
field is an integer specifying the time a message or reaction was sent.
For incoming messages, the
.Em second
field is an integer specifying the time a message or reaction was received.
The times in these two fields are specified in Unix time, with millisecond
precision.
.Pp
The
.Em third
field is an integer specifying the thread id.
.Pp
The
.Em fourth
field is an integer specifying the message type.
A value of 0 indicates an incoming message, 1 indicates an outgoing message,
and 2 indicates a reaction.
Reaction records always immediately follow after the record of the message they
are reactions to.
.Pp
The
.Em fifth
field is an integer specifying the number of attachments a message has.
For reactions, this field is 0.
.Pp
The
.Em sixth
field is a string containing the phone number of the sender or recipient.
For outgoing messages sent to a group, this field is the string
.Sq group .
.Pp
The
.Em seventh
field is a string containing the name of the sender, recipient or group.
.Pp
The
.Em eighth
field is a string containing the message text or reaction emoji.
.Sh EXIT STATUS
.Ex -std
.Sh EXAMPLES
Export all messages from the file
.Pa signal.backup
to the directory
.Pa messages :
.Pp
.Dl $ sigbak export-messages signal.backup messages
.Pp
First create a passphrase file, then export all messages and attachments:
.Bd -literal -offset indent
$ echo 01234 56789 01234 56789 01234 56789 > passfile
$ sigbak export-messages -p passfile signal.backup messages
$ sigbak export-attachments -p passfile signal.backup attachments
.Ed
.Pp
Export the SQLite database and use
.Xr sqlite3 1
to view the contents of the
.Sq message
table:
.Bd -literal -offset indent
$ sigbak export-database signal.backup signal.db
$ sqlite3 signal.db 'select * from message' | less
.Ed
.Sh SEE ALSO
.Lk https://github.com/tbvdm/sigbak
.Sh AUTHORS
The
.Nm
utility was written by
.An Tim van der Molen Aq Mt tim@kariliq.nl .


sigbak.c
/*
 * Copyright (c) 2018 Tim van der Molen <tim@kariliq.nl>
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

#include "config.h"

#include <ctype.h>
#include <fcntl.h>
#include <libgen.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include "sigbak.h"

extern const struct cmd_entry cmd_check_backup_entry;
extern const struct cmd_entry cmd_dump_backup_entry;
extern const struct cmd_entry cmd_export_attachments_entry;
extern const struct cmd_entry cmd_export_avatars_entry;
extern const struct cmd_entry cmd_export_database_entry;
extern const struct cmd_entry cmd_export_messages_entry;
extern const struct cmd_entry cmd_export_stickers_entry;

static const struct cmd_entry *commands[] = {
	&cmd_check_backup_entry,
	&cmd_dump_backup_entry,
	&cmd_export_attachments_entry,
	&cmd_export_avatars_entry,
	&cmd_export_database_entry,
	&cmd_export_messages_entry,
	&cmd_export_stickers_entry
};

__dead static void
usage(const char *cmd, const char *args)
{
	fprintf(stderr, "usage: %s %s %s\n", getprogname(), cmd, args);
	exit(1);
}

int
get_passphrase(const char *passfile, char *buf, size_t bufsize)
{
	int	 fd;
	ssize_t	 len;
	char	*c, *d;

	if (bufsize == 0)
		return -1;

	if (passfile == NULL) {
		if (readpassphrase("Enter 30-digit passphrase (whitespace is "
		    "ignored): ", buf, bufsize, 0) == NULL) {
			warnx("Cannot read passphrase");
			explicit_bzero(buf, bufsize);
			return -1;
		}
	} else {
		if (strcmp(passfile, "-") == 0)
			fd = STDIN_FILENO;
		else if ((fd = open(passfile, O_RDONLY)) == -1) {
			warn("%s", passfile);
			return -1;
		}

		if ((len = read(fd, buf, bufsize - 1)) == -1) {
			warn("%s", passfile);
			explicit_bzero(buf, bufsize);
			close(fd);
			return -1;
		}

		if ((c = memchr(buf, '\n', len)) != NULL)
			len = c - buf;

		buf[len] = '\0';

		if (fd != STDIN_FILENO)
			close(fd);
	}

	/* Remove whitespace */
	for (c = d = buf; *c != '\0'; c++)
		if (!isspace((unsigned char)*c))
			*d++ = *c;
	*d = '\0';

	return 0;
}

int
unveil_dirname(const char *path, const char *perms)
{
	char *dir, *tmp;

	if ((tmp = strdup(path)) == NULL) {
		warn(NULL);
		return -1;
	}

	if ((dir = dirname(tmp)) == NULL) {
		warnx("dirname() failed");
		free(tmp);
		return -1;
	}

	if (unveil(dir, perms) == -1) {
		warn("unveil: %s", dir);
		free(tmp);
		return -1;
	}

	free(tmp);
	return 0;
}

void
sanitise_filename(char *name)
{
	char *c;

	if (strcmp(name, ".") == 0) {
		name[0] = '_';
		return;
	}

	if (strcmp(name, "..") == 0) {
		name[0] = name[1] = '_';
		return;
	}

	for (c = name; *c != '\0'; c++)
		if (*c == '/' || iscntrl((unsigned char)*c))
			*c = '_';
}

char *
get_recipient_filename(struct sbk_recipient *rcp, const char *ext)
{
	char		*fname;
	const char	*detail, *name;
	int		 ret;

	name = sbk_get_recipient_display_name(rcp);

	if (rcp->type == SBK_GROUP)
		detail = "group";
	else
		detail = rcp->contact->phone;

	if (ext == NULL)
		ext = "";

	if (detail != NULL)
		ret = asprintf(&fname, "%s (%s)%s", name, detail, ext);
	else
		ret = asprintf(&fname, "%s%s", name, ext);

	if (ret == -1) {
		warnx("asprintf() failed");
		return NULL;
	}

	sanitise_filename(fname);
	return fname;
}

int
main(int argc, char **argv)
{
	const struct cmd_entry	*cmd;
	size_t			 i;

	setprogname(argv[0]);

	if (argc < 2)
		usage("command", "[argument ...]");

	argc--;
	argv++;
	cmd = NULL;

	for (i = 0; i < nitems(commands); i++)
		if (strcmp(argv[0], commands[i]->name) == 0 ||
		    strcmp(argv[0], commands[i]->alias) == 0) {
			cmd = commands[i];
			break;
		}

	if (cmd == NULL)
		errx(1, "%s: Invalid command", argv[0]);

	switch (cmd->exec(argc, argv)) {
	case CMD_OK:
		return 0;
	case CMD_ERROR:
		return 1;
	case CMD_USAGE:
		usage(cmd->name, cmd->usage);
		return 1;
	}
}
